src/realm/transaction.cpp

```diff
@@ -902,4 +907,100 @@ void Transaction::set_transact_stage(DB::TransactStage stage) noexcept
     m_transact_stage = stage;
 }
 
+class NodeTree {
+public:
+    NodeTree(size_t evac_limit, size_t& work_limit)
+        : m_evac_limit(evac_limit)
+        , m_work_limit(work_limit)
+        , m_moved(0)
+    {
+    }
+    ~NodeTree()
+    {
+        // std::cout << "Moved: " << m_moved << std::endl;
+    }
+
+    /// Function used to traverse the node tree and "copy on write" nodes
+    /// that are found above the evac_limit. The function will return
+    /// when either the whole tree has been travesed or when the work_limit
+    /// has been reached.
+    /// \param current_node - node to process.
+    /// \param level - the level at which current_node is placed in the tree
+    /// \param progress - When the traversal is initiated, this vector identifies at which
+    ///                   node the process should be resumed. It is subesequently updated
+    ///                   to point to the node we have just processed
+    bool trv(Array& current_node, unsigned level, std::vector<size_t>& progress)
+    {
+        if (current_node.is_read_only()) {
+            auto byte_size = current_node.get_byte_size();
+            if ((current_node.get_ref() + byte_size) > m_evac_limit) {
+                current_node.copy_on_write();
+                m_moved++;
+                if (m_work_limit > byte_size) {
+                    m_work_limit -= byte_size;
+                }
+                else {
+                    m_work_limit = 0;
+                    return false;
+                }
+            }
+        }
+
+        if (current_node.has_refs()) {
+            auto sz = current_node.size();
+            if (progress.size() == level) {
+                progress.push_back(0);
+            }
+            size_t ndx = progress[level];
+            while (ndx < sz) {
+                auto val = current_node.get(ndx);
+                if (val && !(val & 1)) {
+                    Array arr(current_node.get_alloc());
+                    arr.set_parent(&current_node, ndx);
+                    arr.init_from_parent();
+                    if (!trv(arr, level + 1, progress)) {
```

Comment:
There is a risk of stack overflow when using recursion to a depth proportional to a user's data, what do you think about rewriting this into an iterative algorithm?

Reply:
The depth only grows with log(n), so even very large tables will have limited depth. Being able to recuse down the cluster tree is used in a lot of places.